### PR TITLE
3/465 refactoring lootbox and creating more concrete tests

### LINKED ISSUE #465: Refactoring LootBox and Creating More Concrete Tests
# **Description**
**Task:** Refactoring LootBox and Improving TestCases (#3).
**Feature:** Item/Inventory Sprint 3 features (#322) 

This task involves refactoring the lootbox system by removing the EarlyGameLootBox, MidGameLootBox, and LateGameLootBox classes, as well as their respective loot tables (EarlyGameLootTable, MidGameLootTable, LateGameLootTable). These will be replaced with a more flexible approach using a loottables.json configuration file and the LootboxFactory. The configuration file will contain the loot tables, number of rolls, and other necessary properties to dynamically generate different types of lootboxes without requiring separate classes for each one. The LootboxFactory will be refactored to read from the loottables.json file and create lootboxes based on its contents. Test cases will also be updated to reflect the new structure.
 

# **Milestones**
List of steps that need to be completed for this task.
- [ ] Remove EarlyGameLootBox, MidGameLootBox, and LateGameLootBox classes from the codebase (Oct 2)
- [ ] Remove their respective loot tables (EarlyGameLootTable, MidGameLootTable, LateGameLootTable)  (Oct 2)
- [ ] Create the LootboxFactory to dynamically create lootboxes based on the loottables.json file (Oct 2)
- [ ] Update the loottables.json file to include configuration for different lootbox types, defining the loot table, number of rolls, description, and image path for each box (Oct 2)
- [ ]  Update and expand test cases to ensure lootbox creation is handled dynamically via the factory and JSON configuration (Oct 2)
- [ ] Validate that all tests work with the new generalized lootbox system, removing dependencies on specific lootbox classes (Oct 2)
 
**Completion Deadline:** Oct 2

# **Documentation**
- TDB
 
# **Member**
- Lewis Hickson (lhick2108) (prehistoricmesopotamiantablet)

### PR DESCRIPTION
**Pull request by lhick2108**
# Description

This PR refactors the LootBoxFactory to dynamically generate loot boxes based on a JSON configuration file (loottables.json). The refactor allows for the removal of specific loot box classes such as EarlyGameLootBox, MidGameLootBox, and LateGameLootBox, along with their associated loot tables, simplifying the codebase. The factory now uses reflection to create loot tables and items, improving flexibility and maintainability. Additionally, the new implementation supports constructor argument parsing for dynamic item creation.

The main changes include:

Refactored LootBoxFactory to load loot boxes from loottables.json.
Removed hardcoded loot box classes and loot tables.
Added dynamic item creation based on the JSON configuration.
Updated tests to ensure proper coverage of the new loot box generation process.

Fixes / Closes #465 

## Type of change

- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

The changes have been tested through:

- Unit Tests: New and updated tests were added to verify that loot boxes are correctly created from the JSON file and that all loot table logic is working as expected.

- Manual Testing: The game was run locally to ensure that the new loot box system behaves correctly, and that items are correctly added to the player's inventory or dropped on the ground if the inventory is full.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: source/core/src/main/com/csse3200/game/components/lootboxview/LootBoxOverlayComponent.java [hunk 1]
```diff
@@ -59,20 +60,32 @@ private void showLoot(List<AbstractItem> items) {
         // Clear the previous contents of the overlay
         overlayWindow.clearChildren();
 
-        // Add new items to the overlay
+        // Create a new table for displaying the loot items
         Table contentTable = new Table(skin);
+
+        // Set up a grid-like layout to prevent cramming, adjust columns based on items
+        int numColumns = 4;  // Adjust the number of columns to suit your layout
```
Comment (Jul-Wong) [inline source/core/src/main/com/csse3200/game/components/lootboxview/LootBoxOverlayComponent.java:67]: I wonder if you could move these up into a private final int numColumns = 4, so it's easier to identify and change without having to go over the whole class.

**Review by Jul-Wong — COMMENTED** [review]
Looks mostly good!

File: source/core/src/main/com/csse3200/game/entities/factories/LootBoxFactory.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+        for (JsonValue lootBoxJson : root.get("lootBoxes")) {
+            if (lootBoxJson.getString("name").equals(lootBoxName)) {
+                BaseLootTable lootTable = createLootTable(lootBoxJson);
+                int rolls = lootBoxJson.getInt("rolls");
+                String description = lootBoxJson.getString("description");
+                String imagePath = lootBoxJson.getString("image");
+                int code = lootBoxJson.getInt("code");
+
+                // Create the loot box
+                UniversalLootBox lootBox = new UniversalLootBox(lootTable, rolls, player, description, imagePath, code);
```
Comment (EllaBerglas) [inline source/core/src/main/com/csse3200/game/entities/factories/LootBoxFactory.java:48]: the variable is redundant (pops up as a warning)

File: source/core/src/main/com/csse3200/game/entities/factories/LootBoxFactory.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+            double weight = itemJson.getDouble("weight");
+
+            // Parse constructor arguments and their types
+            Object[][] parsedData = parseConstructorArgsAndTypes(itemJson.get("constructorArgs"));
+            Object[] constructorArgs = parsedData[0];  // Extract the arguments
+            Class<?>[] paramTypes = (Class<?>[]) parsedData[1];  // Extract the parameter types
+
+            // Try to create the item using reflection and add it to the loot table
+            try {
+                AbstractItem item = createItem(itemClass, paramTypes, constructorArgs);
```
Comment (EllaBerglas) [inline source/core/src/main/com/csse3200/game/entities/factories/LootBoxFactory.java:78]: is this item ever used?

**Review by EllaBerglas — APPROVED** [review]
A few warnings you can fix up but otherwise looks good

**Pull request merged**